File: Loenn/triggers/effect_trigger.lua
local effectTrigger = {}

effectTrigger.name = "Ingeste/EffectTrigger"
effectTrigger.placements = {
    {
        name = "particle",
        data = {
            width = 16,
            height = 16,
            effectType = "sparkles",
            intensity = 1.0,
            duration = 3.0,
            triggerOnce = true
        }
    },
    {
        name = "screen_shake",
        data = {
            width = 24,
            height = 24,
            effectType = "screen_shake",
            intensity = 0.5,
            duration = 1.0,
            triggerOnce = false
        }
    },
    {
        name = "color_grade",
        data = {
            width = 32,
            height = 32,
            effectType = "color_grade",
            intensity = 0.8,
            duration = 2.0,
            triggerOnce = false
        }
    }
}

effectTrigger.fieldInformation = {
    effectType = {
        options = {"sparkles", "screen_shake", "color_grade", "wind", "lightning"},
        editable = false
    },
    intensity = {
        fieldType = "number",
        minimumValue = 0.1,
        maximumValue = 2.0
    },
    duration = {
        fieldType = "number",
        minimumValue = 0.5,
        maximumValue = 10.0
    },
    triggerOnce = {
        fieldType = "boolean"
    }
}

function effectTrigger.sprite(room, entity)
    local width = entity.width or 16
    local height = entity.height or 16
    local colors = {
        sparkles = {1.0, 0.9, 0.3, 0.7},
        screen_shake = {0.9, 0.3, 0.3, 0.7},
        color_grade = {0.6, 0.3, 0.9, 0.7},
        wind = {0.3, 0.9, 0.9, 0.7},
        lightning = {0.9, 0.9, 0.9, 0.8}
    }
    
    local color = colors[entity.effectType] or colors.sparkles
    
    return {
        {
            texture = "ahorn/entityTrigger",
            x = entity.x,
            y = entity.y,
            scaleX = width / 8,
            scaleY = height / 8,
            color = color
        }
    }
end

function effectTrigger.selection(room, entity)
    local width = entity.width or 16
    local height = entity.height or 16
    return {entity.x, entity.y, width, height}
end

return effectTrigger

File: Source/Core/Triggers/EffectTrigger.cs
namespace DesoloZantas.Core.Core.Triggers {
    /// <summary>
    /// Trigger for various visual effects like particles, screen shake, etc.
    /// </summary>
    [CustomEntity("Ingeste/EffectTrigger")]
    public class EffectTrigger : Trigger {
        public enum EffectType {
            Sparkles,
            ScreenShake,
            ColorGrade,
            Wind,
            Lightning
        }

        private EffectType effectType;
        private float intensity;
        private float duration;
        private bool triggerOnce;
        private bool hasTriggered;
        private Level level;

        public EffectTrigger(EntityData data, Vector2 offset) : base(data, offset) {
            string effectName = data.Attr(nameof(effectType), "sparkles");
            if (!System.Enum.TryParse(effectName, true, out effectType)) {
                effectType = EffectType.Sparkles;
            }

            intensity = data.Float(nameof(intensity), 1.0f);
            duration = data.Float(nameof(duration), 3.0f);
            triggerOnce = data.Bool(nameof(triggerOnce), true);
            hasTriggered = false;
        }

        public override void Added(Scene scene) {
            base.Added(scene);
            level = scene as Level;
        }

        public override void OnEnter(global::Celeste.Player player) {
            base.OnEnter(player);

            if (!triggerOnce || !hasTriggered) {
                triggerEffect();
                hasTriggered = true;
            }
        }

        private void triggerEffect() {
            if (level == null) return;

            switch (effectType) {
                case EffectType.Sparkles:
                    createSparkleEffect();
                    break;
                case EffectType.ScreenShake:
                    createScreenShake();
                    break;
                case EffectType.ColorGrade:
                    createColorGradeEffect();
                    break;
                case EffectType.Wind:
                    createWindEffect();
                    break;
                case EffectType.Lightning:
                    createLightningEffect();
                    break;
            }

            // Play appropriate sound
            Audio.Play("event:/game/general/crystalheart_pulse", Position);
        }

        private void createSparkleEffect() {
            for (int i = 0; i < (int)(20 * intensity); i++) {
                Vector2 particlePos = Position + new Vector2(
                    Calc.Random.Range(-Width / 2, Width / 2),
                    Calc.Random.Range(-Height / 2, Height / 2)
                );

                level.ParticlesBG.Emit(ParticleTypes.Dust, particlePos, Color.Gold);
            }
        }

        private void createScreenShake() {
            level.Shake(duration * intensity);
        }

        private void createColorGradeEffect() {
            // This would need custom implementation for color grading
            // For now, just add some visual feedback
            level.Flash(Color.Purple * (intensity * 0.3f));
        }

        private void createWindEffect() {
            // Create wind particles
            for (int i = 0; i < (int)(15 * intensity); i++) {
                Vector2 particlePos = Position + new Vector2(
                    Calc.Random.Range(-Width, Width),
                    Calc.Random.Range(-Height, Height)
                );

                level.ParticlesBG.Emit(ParticleTypes.Dust, particlePos);
            }
        }

        private void createLightningEffect() {
            // Flash effect to simulate lightning
            level.Flash(Color.White * intensity);

            // Add some particle effects
            for (int i = 0; i < (int)(10 * intensity); i++) {
                Vector2 particlePos = Position + new Vector2(
                    Calc.Random.Range(-Width / 2, Width / 2),
                    Calc.Random.Range(-Height / 2, Height / 2)
                );

                level.ParticlesBG.Emit(ParticleTypes.Dust, particlePos, Color.Cyan);
            }
        }
    }
}



